Tests/Client/Custom Assertions/AssertAsync.swift

```diff
@@ -0,0 +1,310 @@
+    ///
+    /// - Parameters:
+    ///   - expression: The expression to evaluate.
+    ///   - timeout: The maximum time the function waits for the expression results to equal.
+    ///   - message: The message to print when the assertion fails.
+    ///
+    /// - Warning: ⚠️ The expression is evaluated repeatedly during the function execution. It should not have
+    ///   any side effects which can affect its result.
+    static func staysTrue(_ expression: @autoclosure @escaping () -> Bool,
+                          timeout: TimeInterval = defaultTimeoutForInversedExpecations,
+                          message: @autoclosure @escaping () -> String = "Failed to stay `TRUE`",
+                          file: StaticString = #file,
+                          line: UInt = #line) -> Assertion {
+        
+        Assertion { assertion, elapsedTime in
+            if elapsedTime >= timeout {
+                // Success
+                assertion.isActive = false
+                
+            } else if expression() == false {
+                // Failure
+                XCTFail(message(), file: file, line: line)
+                assertion.isActive = false
+            }
+        }
+    }
+}
+
+// MARK: - Async runner
+
+/// Allows creating and evaluating asynchronous assertions with synchronous-like syntax. When called, stops the test
+/// execution and waits for the assertions to be fulfilled.
+///
+/// There are two ways of using `AssertAsync`:
+///   - For single assertions, you can use it directly using the convenience static function helpers:
+///   ```
+///   func test() {
+///     // ...
+///     AssertAsync.willBeNil(expression)
+///   }
+///   ```
+///
+///   - If you have multiple assertions you want to evaluate at the same time, you can use the following syntax:
+///   ```
+///   func test() {
+///     // ...
+///     AssertAsync {
+///         Assert.willBeNil(expression1)
+///         Assert.staysFalse(expression2)
+///     }
+///   }
+///   ```
+struct AssertAsync {
+    
+    // This shouldn't be needed and it's just a workaround for https://bugs.swift.org/browse/SR-11628. Remove when possible.
+    @discardableResult
+    init(@AssertionBuilder singleBuilder: () -> Assertion) {
+        self.init(builder: {
+            let built = singleBuilder()
+            return [built]
+        })
+    }
+    
+    @discardableResult
+    init(@AssertionBuilder builder: () -> [Assertion]) {
+        var assertions = builder()
+        let startTimestamp = Date().timeIntervalSince1970
+
+        while assertions.contains(where: { $0.isActive }) {
+            let elapsedTime = Date().timeIntervalSince1970 - startTimestamp
+            for idx in assertions.indices {
+                if assertions[idx].isActive {
+                    assertions[idx].evaluate(elapsedTime: elapsedTime)
+                }
+            }
+            _ = XCTWaiter.wait(for: [.init()], timeout: evaluationPeriod)
```

Comment:
I'm not sure what you mean here. Here we use `XCTWaiter` directly. It doesn't produce failure by default. It returns `XCTWaiter.Result` and you have to call `XCTFail` manually. Here we ignore the result because we use it just for waiting.
You maybe confused it with `wait(for: expactation)` called on `XCTestCase`?

Reply:
Exactly as you've predicted, I confused it. Yet something I didn't know 😅 TIL 👍